rollup/da_syncer/blob_client/blob_client_list.go

```diff
@@ -0,0 +1,65 @@
+package blob_client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/crypto/kzg4844"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/rollup/da_syncer/serrors"
+)
+
+type BlobClientList struct {
+	list   []BlobClient
+	curPos int
+}
+
+func NewBlobClientList(blobClients ...BlobClient) *BlobClientList {
+	return &BlobClientList{
+		list:   blobClients,
+		curPos: 0,
+	}
+}
+
+func (c *BlobClientList) GetBlobByVersionedHashAndBlockNumber(ctx context.Context, versionedHash common.Hash, blockNumber uint64) (*kzg4844.Blob, error) {
+	if len(c.list) == 0 {
+		return nil, fmt.Errorf("BlobClientList.GetBlobByVersionedHash: list of BlobClients is empty")
+	}
+
+	for i := 0; i < len(c.list); i++ {
+		blob, err := c.list[c.curPos].GetBlobByVersionedHashAndBlockNumber(ctx, versionedHash, blockNumber)
+		if err == nil {
+			return blob, nil
+		}
+		c.nextPos()
+		// there was an error, try the next blob client in following iteration
+		log.Warn("BlobClientList: failed to get blob by versioned hash from BlobClient", "err", err, "blob client pos in BlobClientList", c.curPos)
+	}
+
+	// if we iterated over entire list, return a temporary error that will be handled in syncing_pipeline with a backoff and retry
+	return nil, serrors.NewTemporaryError(errors.New("BlobClientList.GetBlobByVersionedHash: failed to get blob by versioned hash from all BlobClients"))
+}
+
+func (c *BlobClientList) nextPos() {
+	c.curPos = (c.curPos + 1) % len(c.list)
+}
+
+func (c *BlobClientList) AddBlobClient(blobClient BlobClient) {
+	c.list = append(c.list, blobClient)
+}
+
+func (c *BlobClientList) RemoveBlobClient(blobClient BlobClient) {
+	c.list = append(c.list, blobClient)
+	for pos, client := range c.list {
+		if client == blobClient {
```

Comment:
what will `RemoveBlobClient` be used (seems not used for the moment)? and for the "==" of interfaces, should we make sure that each type of interface implementation is comparable?

Reply:
Removed in 6d1e5cec498e29c852abf25a05d5c858f722a562